### PR TITLE
Rewrite exotic LAST_VALUE/FIRST_VALUE to self-reference.

### PR DESCRIPTION
**Pull request by kgyrtkirk**
* rewrite `LAST_VALUE(x) OVER (ORDER BY y)` to `LAG(x,0) OVER (ORDER BY y)`
  * the current `LAST_VALUE` logic has no idea about the `frame` itself; and doesn't care about the `order by`
* not directly to `x` because some queries get unplannable that way
* restrict `NTILE` from framing - as its not supported
* add test to ensure that all of the `KNOWN_WINDOW_FNS`'s framing is accounted for

Fixes #XXXX.

#### Release note

<hr>

##### Key changed/added classes in this PR
 * `MyFoo`
 * `OurBar`
 * `TheirBaz`

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java [hunk 1]
```diff
@@ -136,6 +136,24 @@ public void validateWindow(SqlNode windowOrId, SqlValidatorScope scope, @Nullabl
       );
     }
 
+    boolean hasBounds = lowerBound != null || upperBound != null;
```
Comment (soumyava) [inline sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java:139]: Can we add a negative test here to indicate what kind of queries are we restricting so that the user can get an idea. Also good candidate to indicate in the docs as well as NTILE does not support framing

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/CalciteRulesManager.java [hunk 2]
```diff
@@ -330,6 +331,7 @@ private Program buildReductionProgram(final PlannerContext plannerContext, final
       // make it impossible to convert to COALESCE.
       builder.addRuleInstance(new CaseToCoalesceRule());
       builder.addRuleInstance(new CoalesceLookupRule());
+      builder.addRuleInstance(new RewriteFirstValueLastValueRule());
```
Comment (soumyava) [inline sql/src/main/java/org/apache/druid/sql/calcite/planner/CalciteRulesManager.java:334]: It will be good to callout in the description as to what was the need for this new rule. Maybe a simple example with LAST_VALUE and the reason that was not working and how this would rewrite the same query to using a LAG

File: sql/src/main/java/org/apache/druid/sql/calcite/rule/RewriteFirstValueLastValueRule.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        }
+      }
+      return super.visitOver(over);
+    }
+
+    private RexNode rewriteToReferenceCurrentRow(RexOver over)
+    {
+      // could remove `last_value( x ) over ( .... order by y )`
+      // best would be to: return over.getOperands().get(0);
+      // however that make some queries too good
```
Comment (soumyava) [inline sql/src/main/java/org/apache/druid/sql/calcite/rule/RewriteFirstValueLastValueRule.java:98]: What does this comment mean by makig queries too good ?

File: sql/src/main/java/org/apache/druid/sql/calcite/rule/RewriteFirstValueLastValueRule.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+      return makeOver(
+          over,
+          over.getWindow(),
+          SqlStdOperatorTable.LAG,
```
Comment (soumyava) [inline sql/src/main/java/org/apache/druid/sql/calcite/rule/RewriteFirstValueLastValueRule.java:102]: If the operator.getKind() is LAST_VALUE we transform to LAG(x,0) but if the operator.getKind() is FIRST_VALUE do we need to transform to LEAD(x,0) ? Probably a comment can help here

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteWindowQueryTest.java [hunk 5]
```diff
@@ -64,11 +63,7 @@ public class CalciteWindowQueryTest extends BaseCalciteQueryTest
 
   public static final boolean DUMP_ACTUAL_RESULTS = Boolean.parseBoolean(
       System.getProperty("druid.tests.sql.dumpActualResults")
-  );
-
-  static {
-    NullHandling.initializeForTests();
-  }
+  ) || developerIDEdetected();
```
Comment (soumyava) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteWindowQueryTest.java:66]: Why is this change needed ? Is there a special handling only for eclipse ?

**Review by soumyava — COMMENTED** [review]
Made one pass

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java (hunk 1 above)
Comment (kgyrtkirk) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java:139]: note that I don't see any rational use case how someone could utilize `NTILE` with framing  :D

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/CalciteRulesManager.java (hunk 2 above)
Comment (kgyrtkirk) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/CalciteRulesManager.java:334]: added some description as the apidoc of `RewriteFirstValueLastValueRule` - let me know what you think!

File: sql/src/main/java/org/apache/druid/sql/calcite/rule/RewriteFirstValueLastValueRule.java (hunk 3 above)
Comment (kgyrtkirk) [reply sql/src/main/java/org/apache/druid/sql/calcite/rule/RewriteFirstValueLastValueRule.java:98]: the best would be to just:
```
return over.getOperands().get(0);
```

however that will lead to plans which become unplannable as the query changes to be a window query directly on a dataset - and right now that's [not allowed](https://github.com/apache/druid/blob/ef9637eef10fa1127c924ebe0fbabd338f1b0d13/sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java#L1457).

to not alter that for now: I've choosen to retain the Window function in this case but rewrite it to something which just references the *current_row*'s  value via `LAG(x,0)` instead.

**Review by soumyava — APPROVED** [review]
LGTM ! Thanks for putting in the explanation as the javadocs. If not done recently, please merge with master once and then we can merge this in
